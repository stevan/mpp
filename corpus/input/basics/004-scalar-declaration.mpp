# Scalar declarations with numbers
my $x = 10;
my $y = -5;
my $z = 0;
my $pi = 3.14159;
my $large = 1000000;

# Scalar declarations with strings
my $name = "Alice";
my $greeting = "Hello, World!";
my $empty = "";
my $path = "/usr/local/bin";

# Scalar declarations without initialization
my $uninitialized;
my $pending;

# Multiple declarations
my $a = 1;
my $b = 2;
my $c = 3;

# Declarations with expressions
my $sum = 10 + 20;
my $product = 5 * 6;
my $concat = "Hello" . " " . "World";

# Our declarations
our $global = 100;
our $shared = "shared value";
our $package_var;

# State declarations
state $counter = 0;
state $initialized = 1;
state $memo;

# Array declarations
my @array = (1, 2, 3);
my @names = ("Alice", "Bob", "Charlie");
my @mixed = (1, "two", 3.14);
our @global_array = (10, 20, 30);

# Hash declarations
my %hash = (key => "value");
my %config = (host => "localhost", port => 8080);
our %global_hash = (a => 1, b => 2);

# Reference declarations
my $array_ref = [1, 2, 3];
my $hash_ref = {key => "value"};

# Declarations with special variables
my $default = $_;