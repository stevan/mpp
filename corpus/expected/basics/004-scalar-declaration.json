[
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$x"
    },
    "initializer": {
      "type": "Number",
      "value": "10"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$y"
    },
    "initializer": {
      "type": "UnaryOp",
      "operator": "-",
      "operand": {
        "type": "Number",
        "value": "5"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$z"
    },
    "initializer": {
      "type": "Number",
      "value": "0"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$pi"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "Number",
        "value": "3"
      },
      "right": {
        "type": "Number",
        "value": "14159"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$large"
    },
    "initializer": {
      "type": "Number",
      "value": "1000000"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$name"
    },
    "initializer": {
      "type": "String",
      "value": "\"Alice\""
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$greeting"
    },
    "initializer": {
      "type": "String",
      "value": "\"Hello, World!\""
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$empty"
    },
    "initializer": {
      "type": "String",
      "value": "\"\""
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$path"
    },
    "initializer": {
      "type": "String",
      "value": "\"/usr/local/bin\""
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$uninitialized"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$pending"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$a"
    },
    "initializer": {
      "type": "Number",
      "value": "1"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$b"
    },
    "initializer": {
      "type": "Number",
      "value": "2"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$c"
    },
    "initializer": {
      "type": "Number",
      "value": "3"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$sum"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "+",
      "left": {
        "type": "Number",
        "value": "10"
      },
      "right": {
        "type": "Number",
        "value": "20"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$product"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "*",
      "left": {
        "type": "Number",
        "value": "5"
      },
      "right": {
        "type": "Number",
        "value": "6"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$concat"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ".",
      "left": {
        "type": "BinaryOp",
        "operator": ".",
        "left": {
          "type": "String",
          "value": "\"Hello\""
        },
        "right": {
          "type": "String",
          "value": "\" \""
        }
      },
      "right": {
        "type": "String",
        "value": "\"World\""
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "our",
    "variable": {
      "type": "Variable",
      "name": "$global"
    },
    "initializer": {
      "type": "Number",
      "value": "100"
    }
  },
  {
    "type": "Declaration",
    "declarator": "our",
    "variable": {
      "type": "Variable",
      "name": "$shared"
    },
    "initializer": {
      "type": "String",
      "value": "\"shared value\""
    }
  },
  {
    "type": "Declaration",
    "declarator": "our",
    "variable": {
      "type": "Variable",
      "name": "$package_var"
    }
  },
  {
    "type": "Declaration",
    "declarator": "state",
    "variable": {
      "type": "Variable",
      "name": "$counter"
    },
    "initializer": {
      "type": "Number",
      "value": "0"
    }
  },
  {
    "type": "Declaration",
    "declarator": "state",
    "variable": {
      "type": "Variable",
      "name": "$initialized"
    },
    "initializer": {
      "type": "Number",
      "value": "1"
    }
  },
  {
    "type": "Declaration",
    "declarator": "state",
    "variable": {
      "type": "Variable",
      "name": "$memo"
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "@array"
    },
    "initializer": {
      "type": "List",
      "elements": [
        {
          "type": "Number",
          "value": "1"
        },
        {
          "type": "Number",
          "value": "2"
        },
        {
          "type": "Number",
          "value": "3"
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "@names"
    },
    "initializer": {
      "type": "List",
      "elements": [
        {
          "type": "String",
          "value": "\"Alice\""
        },
        {
          "type": "String",
          "value": "\"Bob\""
        },
        {
          "type": "String",
          "value": "\"Charlie\""
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "@mixed"
    },
    "initializer": {
      "type": "List",
      "elements": [
        {
          "type": "Number",
          "value": "1"
        },
        {
          "type": "String",
          "value": "\"two\""
        },
        {
          "type": "BinaryOp",
          "operator": ".",
          "left": {
            "type": "Number",
            "value": "3"
          },
          "right": {
            "type": "Number",
            "value": "14"
          }
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "our",
    "variable": {
      "type": "Variable",
      "name": "@global_array"
    },
    "initializer": {
      "type": "List",
      "elements": [
        {
          "type": "Number",
          "value": "10"
        },
        {
          "type": "Number",
          "value": "20"
        },
        {
          "type": "Number",
          "value": "30"
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "%hash"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "=>",
      "left": {
        "type": "String",
        "value": "key"
      },
      "right": {
        "type": "String",
        "value": "\"value\""
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "%config"
    },
    "initializer": {
      "type": "List",
      "elements": [
        {
          "type": "BinaryOp",
          "operator": "=>",
          "left": {
            "type": "String",
            "value": "host"
          },
          "right": {
            "type": "String",
            "value": "\"localhost\""
          }
        },
        {
          "type": "BinaryOp",
          "operator": "=>",
          "left": {
            "type": "String",
            "value": "port"
          },
          "right": {
            "type": "Number",
            "value": "8080"
          }
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "our",
    "variable": {
      "type": "Variable",
      "name": "%global_hash"
    },
    "initializer": {
      "type": "List",
      "elements": [
        {
          "type": "BinaryOp",
          "operator": "=>",
          "left": {
            "type": "String",
            "value": "a"
          },
          "right": {
            "type": "Number",
            "value": "1"
          }
        },
        {
          "type": "BinaryOp",
          "operator": "=>",
          "left": {
            "type": "String",
            "value": "b"
          },
          "right": {
            "type": "Number",
            "value": "2"
          }
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$array_ref"
    },
    "initializer": {
      "type": "ArrayLiteral",
      "elements": [
        {
          "type": "Number",
          "value": "1"
        },
        {
          "type": "Number",
          "value": "2"
        },
        {
          "type": "Number",
          "value": "3"
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$hash_ref"
    },
    "initializer": {
      "type": "HashLiteral",
      "pairs": [
        {
          "key": {
            "type": "String",
            "value": "key"
          },
          "value": {
            "type": "String",
            "value": "\"value\""
          }
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$default"
    },
    "initializer": {
      "type": "Variable",
      "name": "$_"
    }
  }
]